### ISSUE TITLE
Quirky --max-length in marian-decoder

### ISSUE THREAD

**Issue by alvations** [marian-nmt/marian#206]
When decoding, if a sentence exceeds the `--max-length`, the decoder stops at the sentence before it. It's sort of a gotcha when the decoder file is always short of lines. 

E.g. if there are 3000 lines in the file to be decoded and the 581th line is > 100, `marian-decoder` only outputs 580 lines. 

```
cat ../wmt-data/en-de/test.en.tokenized.truecased.bped | ~/marian/build/marian-decoder \
-c model.npz.best-ce-mean-words.npz.decoder.yml \
-m model.npz.best-ce-mean-words.npz -v vocab.src.yml vocab.trg.yml \
-b 6 --max-length 100 --mini-batch 10 --maxi-batch 1000 \
-d 0 1 2 3 > test.de.bpe.out
```

And the user has to increase to `--max-length 200` to get all lines decoded.

```
cat ../wmt-data/en-de/test.en.tokenized.truecased.bped | ~/marian/build/marian-decoder \
-c model.npz.best-ce-mean-words.npz.decoder.yml \
-m model.npz.best-ce-mean-words.npz -v vocab.src.yml vocab.trg.yml \
-b 6 --max-length 200 --mini-batch 10 --maxi-batch 1000 \
-d 0 1 2 3 > test.de.bpe.out
```

I think it's still okay since the default value for `--max-length` is 1000 but it was sort of a gotcha moment when the decoder file was always falling short of lines =)

**Comment by emjotde** [comment]
Add –max-length-crop this will cut the line at the max length instead of omitting it. 

From: alvations
Sent: Friday, August 24, 2018 3:29 AM
To: marian-nmt/marian
Cc: Subscribed
Subject: [marian-nmt/marian] Quirky --max-length in marian-decoder (#206)

When decoding, if a sentence exceeds the --max-length, the decoder stops at the sentence before it. It's sort of a gotcha when the decoder file is always short of lines.
E.g. if there are 3000 lines in the file to be decoded and the 581th line is > 100, marian-decoder only outputs 580 lines.
cat ../wmt-data/en-de/test.en.tokenized.truecased.bped | ~/marian/build/marian-decoder \
-c model.npz.best-ce-mean-words.npz.decoder.yml \
-m model.npz.best-ce-mean-words.npz -v vocab.src.yml vocab.trg.yml \
-b 6 --max-length 100 --mini-batch 10 --maxi-batch 1000 \
-d 0 1 2 3 > test.de.bpe.out
And the user has to increase to --max-length 200 to get all lines decoded.
cat ../wmt-data/en-de/test.en.tokenized.truecased.bped | ~/marian/build/marian-decoder \
-c model.npz.best-ce-mean-words.npz.decoder.yml \
-m model.npz.best-ce-mean-words.npz -v vocab.src.yml vocab.trg.yml \
-b 6 --max-length 200 --mini-batch 10 --maxi-batch 1000 \
-d 0 1 2 3 > test.de.bpe.out
I think it's still okay since the default value for --max-length is 1000 but it was sort of a gotcha moment when the decoder file was always falling short of lines =)

**Comment by alvations** [comment]
Maybe I should check whether `--max-length-crop` vs `--max-length 100000` makes a big difference. 

Regardlessly, I guess omitting and returning `\n` is okay but stopping the decoding on the line is sort of quirky, esp. when the no. of lines in the output file is different from the input.

**Comment by emjotde** [comment]
Yeah, the reason for this is that the decoder is using the same input method as training and removing too long lines makes sense there. I can see your point, though.
 
Maybe we should enable cropping by default, but still output a warning during decoding?

**Comment by emjotde** [comment]
@snukky Opinions?

**Comment by snukky** [comment]
An alternative to enabling `--max-length-crop` by default might be generating an empty translation if the decoder sees an input sentence exceeding `--max-length`, and I thought that it already does that.

I'm not sure what is better but in either case marian-decoder should produce a clear warning.

**Comment by frankseide** [comment]
Why is this needed for offline decoding anyway? How about just keeping the option in training, but not automatically propagating it to the decoder, and defaulting to SIZE_MAX in decoding (and fail if out of memory)?

**Comment by alvations** [comment]
BTW, what's the default `SIZE_MAX`?

**Comment by kpu** [comment]
"(and fail if out of memory)"

**Comment by emjotde** [comment]
The usecase is for instance back-translation. It's a bit annoying to have that fail if you have stray long inputs around sentence 2 million.

**Comment by kpu** [comment]
The dominant use case is translating things.  Where you want one line in and one line out.  You also want that for backtranslation.  

https://github.com/kpu/preprocess/blob/master/preprocess/remove_long_lines_main.cc

**Comment by emjotde** [comment]
Can you condense your preference into a paragraph?

**Comment by kpu** [comment]
Marian should preserve the principle of one line in, one line out.  If you're backtranslating web data, the long line should have been removed from the input; there's nothing consistent Marian can do with it.

**Comment by emjotde** [comment]
So your default would be do nothing for translation, same as Frank proposed. Fair enough. 

We can leave `--max-length` and `--max-length-crop` in there anyway.

**Comment by kpu** [comment]
My thumbs up on Frank's comment stands.

**Comment by emjotde** [comment]
OK. @snukky can you take care of this according to the will of the vocal majority?

**Comment by emjotde** [comment]
Should we warn even if there is no enforced length limit?

**Comment by frankseide** [comment]
I would go by what the dominant use case is. If that is translation, there should be a warning if a limit is set (!), as unnoticed truncation would affect your BELU score. If it is back-translation, there should be a warning if not set. Is there any heuristic for discovering whether an invocation is back-translation?

"Marian should preserve the principle of one line in, one line out." Absolutely, this should be non-negotiable. The entire data-file organization rests on the assumption that parallel data files are aligned line by line ("struct of arrays"). In our statistical world, off-by-one alignment errors towards the end of back-translated training data may cause subtle degradations that may go unnoticed (I've seen this once in my work on speech recognition).

"alternative behavior like a blank line" This would require Marian to special-case blank lines when reading back the back-translated input in the subsequent training. Are blank lines valid input under any circumstance? If it is, then one would have to pre-filter for blank lines, which is not much different than pre-filtering for too long input (both require an additional processing step, and to manage an additional version of the data). Instead of a blank line, we could consider writing out a special tag that is understood by Marian's reader (but other tools would not understand it).

**Comment by kpu** [comment]
I vote for no special handling of blank lines.  
![image](https://user-images.githubusercontent.com/247512/44605393-0d743580-a7e1-11e8-98ce-d87fe22e3826.png)

**Comment by emjotde** [comment]
I think blank lines are currently also not handled correctly during translation. They are ignored during training and probably also during translation. Another thing to be fixed. I agree they should just be handed through. 

As for use-cases, back-translation is just translation, no way to identify.

**Comment by frankseide** [comment]
Do we know the corpus size in advance? Then we could say, e.g., translating more than 10,000 sentences is likely back-translation, and then enable a warning.

**Comment by emjotde** [comment]
Not when reading from STDIN.

**Comment by kpu** [comment]
I still contend that there's nothing different in these cases.  If you're feeding long sentences into backtranslation, it's your fault for not cleaning it up before.  Because otherwise you'll have to clean it up after which is messier.

**Comment by emjotde** [comment]
Yes, I can see this as the default behavior. Just asking if we should warn somehow?

**Comment by kpu** [comment]
I would support warning once per process.

**Comment by frankseide** [comment]
I am leaning against. Isn't setting a max length for back-translation something you forget once and never again?

**Comment by kpu** [comment]
@frankseide I'm worried about a user who remembers to set a max length, passes `--max-length` and is confused why it doesn't do anything.

**Comment by frankseide** [comment]
In my mind, if one says `--max-length`, it should do something; but if it is not set, it should default to SIZE_MAX i.e. do nothing. And then Marcin will forget it once in back-translation, waste a few hours, and never forget it again...

**Comment by emjotde** [comment]
I will also pass on all angry issue to you :)

**Comment by frankseide** [comment]
Hey, that's the difference between Windows and UNIX. Windows: What you see is what you get. UNIX: You asked for it, you got it.

**Comment by alvations** [comment]
@emjotde @snukky seems like the same is happening for `marian` (trainer, bda9b18b6cede63b0476e9c144da3f62f03515b1) where the decoder seemed to validation set stopped abruptly when it sees a sentence with max-length > than the default, if `--max-length` is left unset.

**Comment by emjotde** [comment]
Just clean your validation data, dammit :)

**Comment by alvations** [comment]
Yeah, can't really throw them away because it's throwing away 30% of my validation set. 

The default `max-length` assumption needs to be larger, esp. when the input is non-European, no. of tokens is in the range of 500-1000 after BPE/SPM =)

**Comment by emjotde** [comment]
How in the world are >1000 tokens lines useful data?

**Comment by alvations** [comment]
Hohoho, doc-level mess, e.g. single line input:

> ▁以下は 、 前 問 で 触れた ▁ 代替 パワー トレイン 技術 ▁ の説明 です 。 ▁それぞれの 技術 の説明 をよく 読んで から 、 質問 に お答え 下さい 。 ▁ クリーン ディーゼルエンジン : ▁ クリーン ディーゼルエンジン は 、 従来の ガソリンエンジン と同様の 内 燃 機関 を備え ています が 、 燃料 に ガソリン ではなく ▁ 軽油 を使う 点 のみ が異なり ます 。 新技術 の採用 により 、 ディーゼルエンジン の 燃費 性能 の良さ を活かし しつつ 、 排出 ガス は ガソリンエンジン と ▁ 同等 レベル に 低下 させる ことが可能 となっています 。 ▁ ターボ チャージ の ガソリンエンジン : ▁ ターボ チャージ エンジンは 、 燃焼 室 へ 、 より多くの 空気 を送り 込むこと で 排気量 と 熱 効率 を高め ます 。 ▁これにより 自動車の 動力 性能 が高まり 、 燃費 効率 も 向上 します 。 急 加速 した際には ターボ ラグ が発生し 、 稀に 違和感 が 感じられる ことがあります 。 ▁ バッテリー 式 電気自動車 ( B EV ▁/ ▁ EV ): ▁電気 モーター が 常に 自動車の 動力 源 となります 。 自宅 に設置された 充電 設備 から ( または 、 時間は かかる ものの ▁一般的な 住宅 用 電源 から でも ) バッテリー を 充電 できる の で ガソリン を 給油 する必要 は全く なくなります 。 バッテリー 式 電気自動車 は通常 、 非常に 静か です 。 ▁ ハイブリッド車 ( HE V ▁/ ▁ HV ): ▁電気 モーター が ガソリンエンジン を 補助 しながら 機能 するので 燃費 が向上し ます 。 ▁低 速 の場合 、 一定の 範囲 ( 3 km 程度 ) は 電気 モーター だけで 自動車 を動かし ます 。 加速 時 や 高速運転 時 など 、 必要に応じて ガソリンエンジン が 起動 します 。 ▁ ハイブリッド車 は通常 、 停車 中 と 低 速 時に 非常に 静か です 。 また 、 コンセント に つないで 充電 する 必要はありません 。 ▁ プラグイン ハイブリッド車 ( PH EV ▁/ ▁P HV ): ▁ プラグイン ハイブリッド車 は 、 ハイブリッド車 と 似 ています が 、 自宅 に設置された 充電 設備 から ▁( または 、 時間は かかる ものの 一般的な 住宅 用 電源 から ) も バッテリー を 充電 できる 点 のみ が異なり ます 。 これにより ハイブリッド車 に比べ 、 より 長い間 、 ▁電気 だけで 運転 することができ 、 その分 燃費 が向上し ます 。 高速運転 時は 、 バッテリー が 補助 的に 機能 し ガソリン 消費量 を減らす ことができます 。 ▁ バッテリー が空 になった 状態では ハイブリッド車 と同じように 機能 します 。 プラグイン ハイブリッド車 は通常 、 停車 中 と 低 速 時に 非常に 静か です 。 ▁ 天然ガス 自動車 ( C NG ): ▁従来の ガソリンエンジン と似た エンジン を使って います が 、 化石燃料 ではなく 、 より クリーン な 圧縮 天然ガス ( C NG ) ▁また は 液化 天然ガス ( LNG ) を使用している 点が 違います 。 ▁ 燃料電池 自動車 ( FC EV ▁/ ▁F CV ): ▁ 燃料電池 自動車 は 、 電気 をつくり だ すために 燃料電池 を使う 種類の 自動車 で 、 車内 に搭載された 電気 モーター を使って 起動 します 。 ▁ 燃料電池 は 、 ( 車内 に 備蓄 された ) 水素 と 、 空気中 の 酸素 を使って 電気 をつくり 出す の で 水蒸気 のみが 排出 されます 。 燃料電池 自動車 は通常 、 非常に 静か です 

Otherwise, I had to do 

```
cat valid.ja | awk '{ print length, $0 }' | sort -n -s | cut -d" " -f2- > valid.ja.clean
```

**Comment by emjotde** [comment]
Yes. Emphasize on *useful*. This had loads of full-stops in there, why not split? I am pretty sure you are hurting yourself with that.

**Comment by alvations** [comment]
Because no. of sentences in source paragraph != no. of sentences in target paragraph. And non-monotonic sentence alignments between the sentences inside the paragraph =(

I guess that's another pre-processing sentence-alignment step that's necessary.

**Comment by emjotde** [comment]
Definitely. Otherwise garbage in, garbage out.
